examples/wave.cxx

```diff
@@ -0,0 +1,372 @@
+
+
+typedef struct grid{
+  double ox,dx;
+  int nx;
+}grid_s;
+
+
+template <typename T>
+T* allocate(RAJA::Index_type size) {
+
+  T* ptr;
+#if defined(RAJA_ENABLE_CUDA)
+  cudaMallocManaged((void**)&ptr, sizeof(T)*size,cudaMemAttachGlobal);
+#else
+  ptr = new T[size];
+#endif
+  return ptr;
+}
+
+template <typename T>
+void deallocate(T* &ptr) {
+  if (ptr) {
+#if defined(RAJA_ENABLE_CUDA)
+    cudaFree(ptr);
+#else
+    delete[] ptr;
+#endif
+    ptr = nullptr;
+  }
+}
+
+
+
+struct DefaultCPP{};
+
+//Generic RAJA implementation for CPU
+template <typename Policy1, typename Policy2>
+struct Wave {
+  template <typename T>  
+    void operator()(T* P1, T* P2, int nx, double ct) {
+
+    RAJA::forallN< RAJA::NestedPolicy<
+    RAJA::ExecList<RAJA::seq_exec,RAJA::seq_exec >>>(
+    RAJA::RangeSegment(0, nx),
+    RAJA::RangeSegment(0, nx),
+    [=](int ty, int tx) {
+    
+    double coeff[5] = {-1.0/12.0,4.0/3.0,-5.0/2.0,4.0/3.0,-1.0/12.0}; //fourth order scheme
+    
+    // loop over points        
+    const int id = tx + ty*nx;
+    double P_old  = P1[id]; 
+    double P_curr = P2[id]; 
+    
+    //Compute laplacian
+    double lap = 0.0;
+    
+    for(int r=-sr; r<=sr; ++r){
+      const int xi  = (tx+r+nx)%nx;
+      const int idx = xi + nx*ty;
+      lap += coeff[r+sr]*P2[idx]; 
+      
+      const int yi  = (ty+r+nx)%nx;
+      const int idy = tx + nx*yi;
+      lap += coeff[r+sr]*P2[idy];
+    }
+    
+    //finite difference stencil
+    P1[id] = 2*P_curr - P_old + ct*lap;
+          
+    }); 
+            
+  }
+};
+
+
+// Generic CPP implementation
+template <>
+struct Wave<DefaultCPP, DefaultCPP> {
+  template <typename T>
+  void operator()(T* P1, T* P2, int nx, double ct) {
+
+    double coeff[5] = {-1.0/12.0,4.0/3.0,-5.0/2.0,4.0/3.0,-1.0/12.0}; //fourth order scheme
+    
+    // loop over points
+    for(int ty=0; ty<nx; ++ty) {
+      for(int tx=0; tx<nx; ++tx) {
+        
+        const int id = tx + ty*nx;
+        T P_old  = P1[id]; 
+        T P_curr = P2[id]; 
+        
+        //Compute laplacian
+        T lap = 0.0;
+        
+        for(int r=-sr; r<=sr; ++r) {
+          const int xi  = (tx+r+nx)%nx;
+          const int idx = xi + nx*ty;
+          lap += coeff[r+sr]*P2[idx]; 
+          
+          const int yi  = (ty+r+nx)%nx;
+          const int idy = tx + nx*yi;
+          lap += coeff[r+sr]*P2[idy];
+        }
+        
+        //finite difference stencil
+        P1[id] = 2*P_curr - P_old + ct*lap;
+      }
+    }
+    
+  }
+
+};
+
+// RAJA implementation for CUDA
+template <size_t BLOCK_SIZE>
+struct Wave<RAJA::cuda_exec<BLOCK_SIZE>, RAJA::cuda_exec<BLOCK_SIZE> > {
+  template <typename T>
+    void operator()(T* P1, T* P2, int nx, double ct) {
+
+    RAJA::forallN<RAJA::NestedPolicy<
+    RAJA::ExecList<
+    RAJA::cuda_threadblock_y_exec<BLOCK_SIZE>,
+    RAJA::cuda_threadblock_x_exec<BLOCK_SIZE> >>>(
+    RAJA::RangeSegment(0, nx), RAJA::RangeSegment(0, nx), [=] __device__ (int tidy, int tidx) {
```

Comment:
Does this only need to be split out because of __device__?  If so, for the sake of examples at least, I would say __host__ __device__ would be worth it to have just one RAJA version in here.